Support ollama

Smpests opened the pull request:
<!--
Thanks for contributing to GraphRAG!

Please do not make *Draft* pull requests, as they still notify anyone watching the repo.

Create a pull request when it is ready for review and feedback.

About this template

The following template aims to help contributors write a good description for their pull requests.
We'd like you to provide a description of the changes in your pull request (i.e. bugs fixed or features added), the motivation behind the changes, and complete the checklist below before opening a pull request.

Feel free to discard it if you need to (e.g. when you just fix a typo). -->

## Description

Supports Ollama, which provides local free large LLM models,for those who need to use local APIs.

## Related Issues

None

## Proposed Changes

1.Add ollama package in graphrag.llm;
2.Add ollama package in graphrag.query.llm;
3.Some details to make the changes works: 
  graphrag.llm.openai.utils.py moved to graphrag.llm.utils.py;
  add new types in graphrag.config.enum.LLMType;
  ...

## Checklist

- [ √] I have tested these changes locally.
- [ √] I have reviewed the code changes.
- [ un] I have updated the documentation (if necessary).
- [ ×] I have added appropriate unit tests (if applicable).

## Additional Notes

Follow https://microsoft.github.io/graphrag/get_started with 24720 chars book.txt，graphrag index and graphrag query were passed.

Part of settings.yaml
llm:
  type: ollama_chat # or azure_openai_chat
  model: llama3.1:8b
  model_supports_json: false # recommended if this is available for your model.
  max_tokens: 12800
  api_base: http://localhost:11434
  concurrent_requests: 2 # the number of parallel inflight requests that may be made
embeddings:
  llm:
    type: ollama_embedding # or azure_openai_embedding
    model: nomic-embed-text:latest
    api_base: http://localhost:11434
    concurrent_requests: 2 # the number of parallel inflight requests that may be made

Smpests commented:
> @Smpests please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement

@microsoft-github-policy-service agree

JoedNgangmeni commented:
I get the following error when running this fork: 

'ollama_chat' is not a valid LLMType

JoedNgangmeni commented:
I've updated my yaml and llm type files but am now getting this error: 
![image](https://github.com/user-attachments/assets/88d22cca-57db-4192-af8c-e23bd46fdf4c)

How do you make sure ollama models are actually being run? I think that is the main issue. 

[stats.json](https://github.com/user-attachments/files/17548177/stats.json)

#####################################################
YAML: 
llm:
  api_key: ${GRAPHRAG_API_KEY}
  type: ollama_chat # or azure_openai_chat
  model: llama3.2:latest
  model_supports_json: true # recommended if this is available for your model.
  # audience: "https://cognitiveservices.azure.com/.default"
  max_tokens: 12800
  # request_timeout: 180.0
  api_base: http://localhost:11434
  # api_version: 2024-02-15-preview
  # organization: <organization_id>
  # deployment_name: <azure_model_deployment_name>
  # tokens_per_minute: 150_000 # set a leaky bucket throttle
  # requests_per_minute: 10_000 # set a leaky bucket throttle
  # max_retries: 10
  # max_retry_wait: 10.0
  # sleep_on_rate_limit_recommendation: true # whether to sleep when azure suggests wait-times
  concurrent_requests: 2 # the number of parallel inflight requests that may be made
  # temperature: 0 # temperature for sampling
  # top_p: 1 # top-p sampling
  # n: 1 # Number of completions to generate

parallelization:
  stagger: 0.3
  # num_threads: 50 # the number of threads to use for parallel processing

async_mode: threaded # or asyncio

embeddings:
  ## parallelization: override the global parallelization settings for embeddings
  async_mode: threaded # or asyncio
  # target: required # or all
  # batch_size: 16 # the number of documents to send in a single request
  # batch_max_tokens: 8191 # the maximum number of tokens to send in a single request
  vector_store:
    type: lancedb
    db_uri: 'output/lancedb'
    collection_name: entity_description_embeddings
    overwrite: true
  # vector_store: # configuration for AI Search
    # type: azure_ai_search
    # url: <ai_search_endpoint>
    # api_key: <api_key> # if not set, will attempt to use managed identity. Expects the `Search Index Data Contributor` RBAC role in this case.
    # audience: <optional> # if using managed identity, the audience to use for the token
    # overwrite: true # or false. Only applicable at index creation time
    # collection_name: <collection_name> # the name of the collection to use
  llm:
    api_key: ${GRAPHRAG_API_KEY}
    type: ollama_embedding # or azure_openai_embedding
    model: mxbai-embed-large:latest
    api_base: http://localhost:11434
    # api_version: 2024-02-15-preview
    # audience: "https://cognitiveservices.azure.com/.default"
    # organization: <organization_id>
    # deployment_name: <azure_model_deployment_name>
    # tokens_per_minute: 150_000 # set a leaky bucket throttle
    # requests_per_minute: 10_000 # set a leaky bucket throttle
    # max_retries: 10
    # max_retry_wait: 10.0
    # sleep_on_rate_limit_recommendation: true # whether to sleep when azure suggests wait-times
    concurrent_requests: 2 # the number of parallel inflight requests that may be made

#####################################################
From Enums.py
class LLMType(str, Enum):
    """LLMType enum class definition."""

    # Embeddings
    OpenAIEmbedding = "openai_embedding"
    AzureOpenAIEmbedding = "azure_openai_embedding"
    OllamaEmbedding = "ollama_embedding"

    # Raw Completion
    OpenAI = "openai"
    AzureOpenAI = "azure_openai"

    # Chat Completion
    OpenAIChat = "openai_chat"
    AzureOpenAIChat = "azure_openai_chat"
    OllamaChat = "ollama_chat"


    # Debug
    StaticResponse = "static_response"

    def __repr__(self):
        """Get a string representation."""
        return f'"{self.value}"'

Smpests commented:
> I've updated my yaml and llm type files but am now getting this error: ![image](https://private-user-images.githubusercontent.com/34903511/380875693-88d22cca-57db-4192-af8c-e23bd46fdf4c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzAxNTQ5NzMsIm5iZiI6MTczMDE1NDY3MywicGF0aCI6Ii8zNDkwMzUxMS8zODA4NzU2OTMtODhkMjJjY2EtNTdkYi00MTkyLWFmOGMtZTIzYmQ0NmZkZjRjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDI4VDIyMzExM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTg1NDE5NDE1MGNjYWFjNmIwNDAzZTU4NjBlMTI0MjlkYjNiMmJhZDg2ODA2MzBkYjFjZjNkYWQxODkxYjFkZjcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.VFMOWh-7E92Ng6K963D1mdxaApGsT-z-eyILZU7F-p4)
> 
> How do you make sure ollama models are actually being run? I think that is the main issue.
> 
> [stats.json](https://github.com/user-attachments/files/17548177/stats.json)
> 
> ```
> #####################################################
> YAML: 
> llm:
>   api_key: ${GRAPHRAG_API_KEY}
>   type: ollama_chat # or azure_openai_chat
>   model: llama3.2:latest
>   model_supports_json: true # recommended if this is available for your model.
>   # audience: "https://cognitiveservices.azure.com/.default"
>   max_tokens: 12800
>   # request_timeout: 180.0
>   api_base: http://localhost:11434
>   # api_version: 2024-02-15-preview
>   # organization: <organization_id>
>   # deployment_name: <azure_model_deployment_name>
>   # tokens_per_minute: 150_000 # set a leaky bucket throttle
>   # requests_per_minute: 10_000 # set a leaky bucket throttle
>   # max_retries: 10
>   # max_retry_wait: 10.0
>   # sleep_on_rate_limit_recommendation: true # whether to sleep when azure suggests wait-times
>   concurrent_requests: 2 # the number of parallel inflight requests that may be made
>   # temperature: 0 # temperature for sampling
>   # top_p: 1 # top-p sampling
>   # n: 1 # Number of completions to generate
> 
> parallelization:
>   stagger: 0.3
>   # num_threads: 50 # the number of threads to use for parallel processing
> 
> async_mode: threaded # or asyncio
> 
> embeddings:
>   ## parallelization: override the global parallelization settings for embeddings
>   async_mode: threaded # or asyncio
>   # target: required # or all
>   # batch_size: 16 # the number of documents to send in a single request
>   # batch_max_tokens: 8191 # the maximum number of tokens to send in a single request
>   vector_store:
>     type: lancedb
>     db_uri: 'output/lancedb'
>     collection_name: entity_description_embeddings
>     overwrite: true
>   # vector_store: # configuration for AI Search
>     # type: azure_ai_search
>     # url: <ai_search_endpoint>
>     # api_key: <api_key> # if not set, will attempt to use managed identity. Expects the `Search Index Data Contributor` RBAC role in this case.
>     # audience: <optional> # if using managed identity, the audience to use for the token
>     # overwrite: true # or false. Only applicable at index creation time
>     # collection_name: <collection_name> # the name of the collection to use
>   llm:
>     api_key: ${GRAPHRAG_API_KEY}
>     type: ollama_embedding # or azure_openai_embedding
>     model: mxbai-embed-large:latest
>     api_base: http://localhost:11434
>     # api_version: 2024-02-15-preview
>     # audience: "https://cognitiveservices.azure.com/.default"
>     # organization: <organization_id>
>     # deployment_name: <azure_model_deployment_name>
>     # tokens_per_minute: 150_000 # set a leaky bucket throttle
>     # requests_per_minute: 10_000 # set a leaky bucket throttle
>     # max_retries: 10
>     # max_retry_wait: 10.0
>     # sleep_on_rate_limit_recommendation: true # whether to sleep when azure suggests wait-times
>     concurrent_requests: 2 # the number of parallel inflight requests that may be made
> ```
> 
> ```
> #####################################################
> From Enums.py
> class LLMType(str, Enum):
>     """LLMType enum class definition."""
> 
>     # Embeddings
>     OpenAIEmbedding = "openai_embedding"
>     AzureOpenAIEmbedding = "azure_openai_embedding"
>     OllamaEmbedding = "ollama_embedding"
> 
>     # Raw Completion
>     OpenAI = "openai"
>     AzureOpenAI = "azure_openai"
> 
>     # Chat Completion
>     OpenAIChat = "openai_chat"
>     AzureOpenAIChat = "azure_openai_chat"
>     OllamaChat = "ollama_chat"
> 
> 
>     # Debug
>     StaticResponse = "static_response"
> 
>     def __repr__(self):
>         """Get a string representation."""
>         return f'"{self.value}"'
> ```

I not only changed the two places you mentioned, you can check **files changed** in this PR, or just use this branch: https://github.com/Smpests/graphrag-ollama/tree/feature/ollama-support.
below is my result:
`graphrag index --root ./ragtest`
![image](https://github.com/user-attachments/assets/676ec4c8-73e2-4e9d-a035-8bbf995ebe72)
`graphrag query --root ./ragtest --method local --query "Who is Scrooge and what are his main relationships?"`
![image](https://github.com/user-attachments/assets/e0117176-7709-4f12-b8af-4d6725d7edbf)

JoedNgangmeni commented:
was the main.py file removed purposefully from graphrag.index? This results in --init and other args not working

JoedNgangmeni commented:
Also, did you encounter any "Error Invoking LLM" errors? Not sure im invoking it right.

Smpests commented:
> Also, did you encounter any "Error Invoking LLM" errors? Not sure im invoking it right.

I didn't.You can try change parallelization.pnum_threads with a lower value in setting.yaml according to your machine, it defaults to 50.

Smpests commented:
> was the main.py file removed purposefully from graphrag.index? This results in --init and other args not working

please check this issue: #1305

Smpests commented:
> was the main.py file removed purposefully from graphrag.index? This results in --init and other args not working

Please checkt this issue: #1305 
Now, graphrag.index = graphrag index

JoedNgangmeni commented:
I'm unsure why it's bugging. 

[logs.json](https://github.com/user-attachments/files/17578910/logs.json)
[indexing-engine.log](https://github.com/user-attachments/files/17578911/indexing-engine.log)

I think the error invoking LLM (even though I set the request timeout in yaml to 12800.0), leads to this models inability to create some reports and summaries. 

The below is from running ```python -m graphrag query --query "who is scrooge?" --root ./ragtest --method global ```

![image](https://github.com/user-attachments/assets/365ea731-066a-4014-bff4-a7be54f8ffd1)

Result from  ```python -m graphrag --query "who is scrooge?" --root ./ragtest --method global ``` :

![image](https://github.com/user-attachments/assets/750dc6e9-a380-4ba0-b7b3-e3240ef2f98c)


Result from  ```python -m graphrag query "who is scrooge?" --root ./ragtest --method global ``` : 

![image](https://github.com/user-attachments/assets/d9f5f407-1214-4e6e-8c8c-30436196a676)

Smpests commented:
> I'm unsure why it's bugging.
> 
> [logs.json](https://github.com/user-attachments/files/17578910/logs.json) [indexing-engine.log](https://github.com/user-attachments/files/17578911/indexing-engine.log)
> 
> I think the error invoking LLM (even though I set the request timeout in yaml to 12800.0), leads to this models inability to create some reports and summaries.
> 
> The below is from running `python -m graphrag query --query "who is scrooge?" --root ./ragtest --method global `
> 
> ![image](https://private-user-images.githubusercontent.com/34903511/381731973-365ea731-066a-4014-bff4-a7be54f8ffd1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA1MTU0OTUsIm5iZiI6MTczMDUxNTE5NSwicGF0aCI6Ii8zNDkwMzUxMS8zODE3MzE5NzMtMzY1ZWE3MzEtMDY2YS00MDE0LWJmZjQtYTdiZTU0ZjhmZmQxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTAyVDAyMzk1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTg2OTc5NWEzODQzYjdkMmE2YjRhNGY2ZWMxMWRkZWY1NTFjN2U1ZjI2ZTVmZmRhZjM2MjU3YmQyODdjYTkxYjAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.5hly4KeSNvjz6MTtyNAPqjwz-A9z1_r8brB9wdPzR2M)
> 
> Result from `python -m graphrag --query "who is scrooge?" --root ./ragtest --method global ` :
> 
> ![image](https://private-user-images.githubusercontent.com/34903511/381732670-750dc6e9-a380-4ba0-b7b3-e3240ef2f98c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA1MTU0OTUsIm5iZiI6MTczMDUxNTE5NSwicGF0aCI6Ii8zNDkwMzUxMS8zODE3MzI2NzAtNzUwZGM2ZTktYTM4MC00YmEwLWI3YjMtZTMyNDBlZjJmOThjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTAyVDAyMzk1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ1M2UzYzEwOTk1MDgxYWRiMzY0ZGMyM2NmOWQ5NmJhMWU2YWY3NmMwNjNjY2RlOTdhYjViNGU1ZDgwYTczYzEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0._atwkZkyB-jD9xy-ZdmVUcBF82uuq1JT1b8IAjYjwfk)
> 
> Result from `python -m graphrag query "who is scrooge?" --root ./ragtest --method global ` :
> 
> ![image](https://private-user-images.githubusercontent.com/34903511/381732883-d9f5f407-1214-4e6e-8c8c-30436196a676.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA1MTU0OTUsIm5iZiI6MTczMDUxNTE5NSwicGF0aCI6Ii8zNDkwMzUxMS8zODE3MzI4ODMtZDlmNWY0MDctMTIxNC00ZTZlLThjOGMtMzA0MzYxOTZhNjc2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTAyVDAyMzk1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNhOWU0ZjY3YmUxZTYxNmJhNjZjODcwYWE2MzRkMGM0OGRmNmNlNjJmNWM3MmQ5YmFhYjAyNzViMjFmYmFjZjgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Rx8iqtysOZgyr_qPqm5Smzo4msd63Yi4doiAEvTAMkI)

Change your query command to `python -m graphrag query --query "who is scrooge?" --root ./ragtest --method global `

I checked your indexing-engine.log, which model did you use？（My case was LLM3.1:8b） Maybe your model response is not a valid json, you should debug step by step or log the model response and see.

I've seen similar error, the model response like: Below is my answer: {"title": "xx"...}, then I added "Answer only JSON, without any other text." to the prompt.

drcrallen commented:
You can get ollama working with just `settings.yaml` with something like this:

```yaml
llm:
  type: openai_chat
  model: qwen2.5:3b-16k
  batch_max_tokens: 8191
  max_tokens: 4000
  api_base: http://127.0.0.1:11434/v1
  max_retries: 3
  model_supports_json: false
  concurrent_requests: 1
```


Where `qwen2.5:3b-16k` is [created ](https://github.com/ollama/ollama?tab=readme-ov-file#customize-a-model) like:

```
FROM qwen2.5:3b-instruct-q6_K
PARAMETER num_ctx 16384
```

For indexing, the problem I've encountered isn't that ollama doesn't have api hooks, it is that the models produce very different results compared to OpenAI's api, many of which are not compatible. I've only gotten qwen2.5 to produce anything reasonable in the models that fit in my (paltry) 8gb card. Looking at this thread it seems like llama3.2 has had luck for people? But are there any others that people have found successful without having to do prompt engineering (aka, take the stock prompts)?

JoedNgangmeni commented:
> > I'm unsure why it's bugging.
> > [logs.json](https://github.com/user-attachments/files/17578910/logs.json) [indexing-engine.log](https://github.com/user-attachments/files/17578911/indexing-engine.log)
> > I think the error invoking LLM (even though I set the request timeout in yaml to 12800.0), leads to this models inability to create some reports and summaries.
> > The below is from running `python -m graphrag query --query "who is scrooge?" --root ./ragtest --method global `
> > ![image](https://private-user-images.githubusercontent.com/34903511/381731973-365ea731-066a-4014-bff4-a7be54f8ffd1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA1MTU0OTUsIm5iZiI6MTczMDUxNTE5NSwicGF0aCI6Ii8zNDkwMzUxMS8zODE3MzE5NzMtMzY1ZWE3MzEtMDY2YS00MDE0LWJmZjQtYTdiZTU0ZjhmZmQxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTAyVDAyMzk1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTg2OTc5NWEzODQzYjdkMmE2YjRhNGY2ZWMxMWRkZWY1NTFjN2U1ZjI2ZTVmZmRhZjM2MjU3YmQyODdjYTkxYjAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.5hly4KeSNvjz6MTtyNAPqjwz-A9z1_r8brB9wdPzR2M)
> > Result from `python -m graphrag --query "who is scrooge?" --root ./ragtest --method global ` :
> > ![image](https://private-user-images.githubusercontent.com/34903511/381732670-750dc6e9-a380-4ba0-b7b3-e3240ef2f98c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA1MTU0OTUsIm5iZiI6MTczMDUxNTE5NSwicGF0aCI6Ii8zNDkwMzUxMS8zODE3MzI2NzAtNzUwZGM2ZTktYTM4MC00YmEwLWI3YjMtZTMyNDBlZjJmOThjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTAyVDAyMzk1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ1M2UzYzEwOTk1MDgxYWRiMzY0ZGMyM2NmOWQ5NmJhMWU2YWY3NmMwNjNjY2RlOTdhYjViNGU1ZDgwYTczYzEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0._atwkZkyB-jD9xy-ZdmVUcBF82uuq1JT1b8IAjYjwfk)
> > Result from `python -m graphrag query "who is scrooge?" --root ./ragtest --method global ` :
> > ![image](https://private-user-images.githubusercontent.com/34903511/381732883-d9f5f407-1214-4e6e-8c8c-30436196a676.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA1MTU0OTUsIm5iZiI6MTczMDUxNTE5NSwicGF0aCI6Ii8zNDkwMzUxMS8zODE3MzI4ODMtZDlmNWY0MDctMTIxNC00ZTZlLThjOGMtMzA0MzYxOTZhNjc2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTAyVDAyMzk1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNhOWU0ZjY3YmUxZTYxNmJhNjZjODcwYWE2MzRkMGM0OGRmNmNlNjJmNWM3MmQ5YmFhYjAyNzViMjFmYmFjZjgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Rx8iqtysOZgyr_qPqm5Smzo4msd63Yi4doiAEvTAMkI)
> 
> Change your query command to `python -m graphrag query --query "who is scrooge?" --root ./ragtest --method global `
> 
> I checked your indexing-engine.log, which model did you use？（My case was LLM3.1:8b） Maybe your model response is not a valid json, you should debug step by step or log the model response and see.
> 
> I've seen similar error, the model response like: Below is my answer: {"title": "xx"...}, then I added "Answer only JSON, without any other text." to the prompt.

does this mean you changed the prompt document from their repo? I'm new to their repo and this LLM space and didn't want to ruin anything. Also, should i make the thing --emit JSON?

Smpests commented:
> > > I'm unsure why it's bugging.
> > > [logs.json](https://github.com/user-attachments/files/17578910/logs.json) [indexing-engine.log](https://github.com/user-attachments/files/17578911/indexing-engine.log)
> > > I think the error invoking LLM (even though I set the request timeout in yaml to 12800.0), leads to this models inability to create some reports and summaries.
> > > The below is from running `python -m graphrag query --query "who is scrooge?" --root ./ragtest --method global `
> > > ![image](https://private-user-images.githubusercontent.com/34903511/381731973-365ea731-066a-4014-bff4-a7be54f8ffd1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA1MTU0OTUsIm5iZiI6MTczMDUxNTE5NSwicGF0aCI6Ii8zNDkwMzUxMS8zODE3MzE5NzMtMzY1ZWE3MzEtMDY2YS00MDE0LWJmZjQtYTdiZTU0ZjhmZmQxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTAyVDAyMzk1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTg2OTc5NWEzODQzYjdkMmE2YjRhNGY2ZWMxMWRkZWY1NTFjN2U1ZjI2ZTVmZmRhZjM2MjU3YmQyODdjYTkxYjAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.5hly4KeSNvjz6MTtyNAPqjwz-A9z1_r8brB9wdPzR2M)
> > > Result from `python -m graphrag --query "who is scrooge?" --root ./ragtest --method global ` :
> > > ![image](https://private-user-images.githubusercontent.com/34903511/381732670-750dc6e9-a380-4ba0-b7b3-e3240ef2f98c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA1MTU0OTUsIm5iZiI6MTczMDUxNTE5NSwicGF0aCI6Ii8zNDkwMzUxMS8zODE3MzI2NzAtNzUwZGM2ZTktYTM4MC00YmEwLWI3YjMtZTMyNDBlZjJmOThjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTAyVDAyMzk1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ1M2UzYzEwOTk1MDgxYWRiMzY0ZGMyM2NmOWQ5NmJhMWU2YWY3NmMwNjNjY2RlOTdhYjViNGU1ZDgwYTczYzEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0._atwkZkyB-jD9xy-ZdmVUcBF82uuq1JT1b8IAjYjwfk)
> > > Result from `python -m graphrag query "who is scrooge?" --root ./ragtest --method global ` :
> > > ![image](https://private-user-images.githubusercontent.com/34903511/381732883-d9f5f407-1214-4e6e-8c8c-30436196a676.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA1MTU0OTUsIm5iZiI6MTczMDUxNTE5NSwicGF0aCI6Ii8zNDkwMzUxMS8zODE3MzI4ODMtZDlmNWY0MDctMTIxNC00ZTZlLThjOGMtMzA0MzYxOTZhNjc2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTAyVDAyMzk1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNhOWU0ZjY3YmUxZTYxNmJhNjZjODcwYWE2MzRkMGM0OGRmNmNlNjJmNWM3MmQ5YmFhYjAyNzViMjFmYmFjZjgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Rx8iqtysOZgyr_qPqm5Smzo4msd63Yi4doiAEvTAMkI)
> > 
> > 
> > Change your query command to `python -m graphrag query --query "who is scrooge?" --root ./ragtest --method global `
> > I checked your indexing-engine.log, which model did you use？（My case was LLM3.1:8b） Maybe your model response is not a valid json, you should debug step by step or log the model response and see.
> > I've seen similar error, the model response like: Below is my answer: {"title": "xx"...}, then I added "Answer only JSON, without any other text." to the prompt.
> 
> I was using ollama 3.2:latest. I will try running it with 3.1:8b.
> 
> Does your response mean you changed the prompt document from their repo? I'm new to their repo and this LLM space and didn't want to ruin anything. I ask because i think the model outputs a parquet file.
> 
> If your answer is yes, should I add --emit json to the prompt?
> 
> ALSO, if we want json outputs how come in your YAM file you set the model support for json to false?

I‘ll try with ollama 3.2:latest, the detail i'll reply later.
Yes, I modified their prompt document.
My setting.yaml set model_supports_json to false, because this parameter make no sense to ollama.

DStarEpoch commented:
Thanks for your sharing, It works well on my local mechine. Will you consider add ollama support for latest version 1.0.0 of graphrag?

sjbae1999 commented:
> > > I'm unsure why it's bugging.
> > > [logs.json](https://github.com/user-attachments/files/17578910/logs.json) [indexing-engine.log](https://github.com/user-attachments/files/17578911/indexing-engine.log)
> > > I think the error invoking LLM (even though I set the request timeout in yaml to 12800.0), leads to this models inability to create some reports and summaries.
> > > The below is from running `python -m graphrag query --query "who is scrooge?" --root ./ragtest --method global `
> > > ![image](https://private-user-images.githubusercontent.com/34903511/381731973-365ea731-066a-4014-bff4-a7be54f8ffd1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA1MTU0OTUsIm5iZiI6MTczMDUxNTE5NSwicGF0aCI6Ii8zNDkwMzUxMS8zODE3MzE5NzMtMzY1ZWE3MzEtMDY2YS00MDE0LWJmZjQtYTdiZTU0ZjhmZmQxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTAyVDAyMzk1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTg2OTc5NWEzODQzYjdkMmE2YjRhNGY2ZWMxMWRkZWY1NTFjN2U1ZjI2ZTVmZmRhZjM2MjU3YmQyODdjYTkxYjAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.5hly4KeSNvjz6MTtyNAPqjwz-A9z1_r8brB9wdPzR2M)
> > > Result from `python -m graphrag --query "who is scrooge?" --root ./ragtest --method global ` :
> > > ![image](https://private-user-images.githubusercontent.com/34903511/381732670-750dc6e9-a380-4ba0-b7b3-e3240ef2f98c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA1MTU0OTUsIm5iZiI6MTczMDUxNTE5NSwicGF0aCI6Ii8zNDkwMzUxMS8zODE3MzI2NzAtNzUwZGM2ZTktYTM4MC00YmEwLWI3YjMtZTMyNDBlZjJmOThjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTAyVDAyMzk1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ1M2UzYzEwOTk1MDgxYWRiMzY0ZGMyM2NmOWQ5NmJhMWU2YWY3NmMwNjNjY2RlOTdhYjViNGU1ZDgwYTczYzEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0._atwkZkyB-jD9xy-ZdmVUcBF82uuq1JT1b8IAjYjwfk)
> > > Result from `python -m graphrag query "who is scrooge?" --root ./ragtest --method global ` :
> > > ![image](https://private-user-images.githubusercontent.com/34903511/381732883-d9f5f407-1214-4e6e-8c8c-30436196a676.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA1MTU0OTUsIm5iZiI6MTczMDUxNTE5NSwicGF0aCI6Ii8zNDkwMzUxMS8zODE3MzI4ODMtZDlmNWY0MDctMTIxNC00ZTZlLThjOGMtMzA0MzYxOTZhNjc2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTAyVDAyMzk1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNhOWU0ZjY3YmUxZTYxNmJhNjZjODcwYWE2MzRkMGM0OGRmNmNlNjJmNWM3MmQ5YmFhYjAyNzViMjFmYmFjZjgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Rx8iqtysOZgyr_qPqm5Smzo4msd63Yi4doiAEvTAMkI)
> > 
> > 
> > Change your query command to `python -m graphrag query --query "who is scrooge?" --root ./ragtest --method global `
> > I checked your indexing-engine.log, which model did you use？（My case was LLM3.1:8b） Maybe your model response is not a valid json, you should debug step by step or log the model response and see.
> > I've seen similar error, the model response like: Below is my answer: {"title": "xx"...}, then I added "Answer only JSON, without any other text." to the prompt.
> 
> I was using ollama 3.2:latest. I will try running it with 3.1:8b.
> 
> Does your response mean you changed the prompt document from their repo? I'm new to their repo and this LLM space and didn't want to ruin anything. I ask because i think the model outputs a parquet file.
> 
> If your answer is yes, should I add --emit json to the prompt?
> 
> ALSO, if we want json outputs how come in your YAM file you set the model support for json to false?

Hello @JoedNgangmeni , I meet the same problem as yours before, but I realize I used 'main' branch of @Smpests repository. Please git clone  this repository & change the branch as 'feature/ollama-support':
https://github.com/Smpests/graphrag-ollama/tree/feature/ollama-support
Then all of your problems will be solved.
Thanks for your effort, @Smpests !

retry51776 commented on graphrag/llm/ollama/create_ollama_client.py:
```diff
@@ -0,0 +1,36 @@
+) -> OllamaClientType:
+    """Create a new Ollama client instance."""
+
+    log.info("Creating OpenAI client base_url=%s", configuration.api_base)
```
Fyi, "Creating OpenAI client base_url=%s" should be "Creating Ollama client base_url=%s"

natoverse commented:
See [new docs](https://microsoft.github.io/graphrag/config/models/) on injecting custom model providers
